PIVX-Labs/MyPIVXWallet
# Masternode and Governance

Coauthored with @Duddino.

Abstract:
Closes #44 and also adds Governance.

What we have done:
- We created a server on our  VPS [link here](https://github.com/Shangaman/PivxNodeController) which relays a pivx node RPCs with simple HTTP requests. Should be fairly easy to self host, and we will provide documentation in the near future.
- We reverse engineered the Pivx Core code and  rewritten all the needed functions in javascript, in particular we added a masternode tab which allows to access an existing masternode or create a new one (the latter is only avaiable on HD wallets, while the former also on single address wallets).
- We created the governance tab which allows users to view and vote on proposals.

How to test:
This PR needs a lot of testing: 
- First of all it must be tested on the test net (if you need tPIVs ask @Duddino  here or on Lab's discord), you should verify that: If you use a single address wallet which contains the collateral (with the corresponding masternode already created) you can import it and eventually enable if it's offline (and also verify that you can vote). Then try to access with a HD wallet and create from scratch a new masternode (and activate it by copying the masternode private key on a vps as usual), then access it and try to vote.
- Same as above with the main net (we couldn't test this part because we don't have enough pivs)
- Verify that MPW properly sanitizes all inputs from a potential rogue node, i.e. that no XSS attacks are possible

Work left to do (deferred to a next PR as this is getting huge):
- Ledger support
- IPv6 and onion address support (Should be fairly straight forward)
- multi masternode support
